### ISSUE TITLE
Postgres Description Sets API: Scoping by community

### ISSUE THREAD

**Issue by siuc-nate** [CredentialEngine/CredentialRegistry#497]
Following up on this comment:
https://github.com/CredentialEngine/CredentialRegistry/issues/471#issuecomment-983605125

We need to be able to scope the `description_sets` endpoint/API by community. For example, we need to be able to get the description set data for a competency framework for the `snhupr` community.

Is something like this available? If so, what is the syntax?

**Comment by siuc-nate** [comment]
We will also need to pass the organization CTID (to check for community access) in a parameter other than `Authorization Token`/`Authorization ApiToken` since that header is currently used for the CE admin key for endpoints like `description_sets`.  Or, if the CE admin key is detected, the community access check needs to be skipped. 

Suggestions? @mparsons-ce @excelsior @edgarf

**Comment by mparsons-ce** [comment]
@neaf 
Yes, I think that it makes sense to scope the description_sets endpoints by community.

**Comment by jeannekitchens** [comment]
Also, the same descriptions sets with the Credential Registry need to be utilized with private communities/registries.

**Comment by neaf** [comment]
@jeannekitchens Do you mean this mapping: https://github.com/CredentialEngine/CredentialRegistry/blob/master/fixtures/description_set_item_maps.yml ?

Asking because in our code Description set is a specific record describing a path for given CTID.

**Comment by jeannekitchens** [comment]
@neaf looks like you made a change that took care of the issue with description sets.  Using the SNHU Credential Finder (it limits display to only data in their private regsitry, I can see the competency frameworks now:

1. https://sandbox.credentialengine.org/snhu/competencyframework/2/SNHU_Competency_Framework and I can click on the details tab on this page, click on the JSON LD icon and see the raw metadata in the SNHU private community.
2. https://sandbox.credentialengine.org/snhu/competencyframework/3/Information_Technologies_(BS) same for this one.

Thank you!

**Comment by neaf** [comment]
@jeannekitchens 

Except of "un-securing" the community I did nothing. Also, even if that did what you expected it works only by coincidence.

I'm also not sure what you mean by "issue with description sets". This task is about introducing a new feature.

**Comment by neaf** [comment]
@siuc-nate To confirm:

1. We'll add `/:community_name/description_sets` endpoint which will return only those that are extracted from envelopes published to given community.
2. `/description_sets` will return `ce_registry` description sets as this is our default community.
3. If the API token validation service returns `valid: true` for the community name specified in the endpoint path, we'll return results. First part of the `Authorization` header will be ignored.

Sounds good?

**Comment by siuc-nate** [comment]
1. This is on the right track, but I suspect that the description set for something published to a private community _may_ also need to reference/include data that comes from one or more public communities (such as `ce-registry`). An example would be a competency framework published to the `snhupr` community that has alignments to competencies published in the `ce-registry` community. If it's possible to include that as well, that would work best. The reverse wouldn't be true though (the description set for something in one community shouldn't include data that comes from a private community unless the user has access to that community).
2. Yes
3. I'm not quite sure what is meant by this. Can you elaborate?

**Comment by siuc-nate** [comment]
It's my understanding that in order to access a resource in a private community, the accessor needs to pass the CTID for their organization in the `Authorization` header, which then gets passed to our accounts system to check whether that organization has access.  Though now that I think about it, that might be the organization's API key (as that would be more secure) rather than the CTID - @mparsons-ce can you clarify?

Anyway, the reason it's an issue is because:
- In order to access something in a private community, the `Authorization` header must be the accessing organization's API key
- In order to access the `description_sets` endpoint, the `Authorization` header must be the CE admin key (different format from the API key)
- In order to access the `description_sets` endpoint inside of a private community, the `Authorization` header must be:
  - The organization's key?
  - The CE admin key?
  - Some concatenation of the two?
  - Pass two `Authorization` headers?

**Comment by neaf** [comment]
(external validation service = `/accountsAPI/Organization/ValidateCommunityAccess`)


I think it's really up to you how you want to treat the endpoint.

Currently we have two kind of endpoints.

a) Normal data access like we use for `/:community_name/{envelopes|resources|search}`
In this case we take the token and validate whether it has access to given community using the external validation service.

b) Administrative access for tasks like publishing, configuration, downloading a ZIP of all envelopes and so on
In this case we're not validating community access in any way and only check whether the token matches administrative user in our database. That means administrative user can manage all communities.

The only exception seems to be V2 Publishing: We take the token from `Authorization` header and use it for finding administrative user in our database AND checking community access with external validation service. Do we expect the same token to work in both places?

**Comment by siuc-nate** [comment]
Maybe we should have two headers, one that is explicitly for use by external parties looking to access data from a private community, and one for us (probably sticking with `Authorization` so we don't break a bunch of our own code) for administrative access.

**Comment by neaf** [comment]
Currently `/description_sets` is an administrative endpoint that requires admin auth token and allows to fetch data from ALL communities.

Newly added `/:community_name/description_sets` endpoint WILL be authenticated with the same admin auth token and will allow to place any community name for fetching data.

If we need to change how the data access is managed here that would be a separate issue as it opens a new set of problems to solve.

**Comment by neaf** [comment]
@siuc-nate 

What I've done:
1. Refactor generation code (fix updates handling)
2. Add reference to a community when generating description set
3. Add `/:community_name/description_sets` endpoint that allows to filter description sets by community
4. Change `/description_sets` endpoint so it returns description sets from default community `cr_registry`

**Regarding 4.** - I've changed it to this because that's the behavior we got in other endpoints like `/envelopes` and `/resources`. Can it be that way or would you rather want to keep the old behavior and return all descriptions sets there?


**Deployment**

Because reference to a community is a new thing on description set model we need to populate it for old data. Given that tomorrow afternoon and during couple of days next week I have a time off (that I forgot about during the meeting) I might not be able to push that forward with deserving attention.

Let me know if you're super blocked with something and I'll try to mitigate.

**Comment by siuc-nate** [comment]
Thanks.

The updated behavior for item 4 is how I thought it worked anyway, so that's a good change (more consistent with the other things, as you point out). We can always setup some other endpoint/parameter if we need to be able to access data regardless of community.

I have a lot on my plate as it is in other areas, so no need to rush the rest for now.

**Comment by siuc-nate** [comment]
I'm attempting to make `POST` requests to https://sandbox.credentialengineregistry.org/snhupr/description_sets but I'm getting a 404. Is that the correct URL?

**Comment by neaf** [comment]
@siuc-nate  As I mentioned in my last comment, I need to prepare some data migration code and be more present during the deployment and backfill. I'll be able to look at it in the next two days and hopefully have that functionality fully working on Monday.

(Anderson you've tagged is not working on that)

**Comment by siuc-nate** [comment]
I think @jeannekitchens added the tag to my post - not sure why. I'll remove it.

**Comment by neaf** [comment]
The APIs should work on sandbox.

**Comment by edgarf** [comment]
👋 @siuc-nate  any luck testing the description sets functionality? I think this is the last action we wait for before @neaf deploys it on production.

**Comment by siuc-nate** [comment]
Sorry, I've been swamped with other stuff.  But I've been digging through some examples on sandbox, and it looks like it works (as far as I can tell anyway).

Go ahead and deploy it.  Thanks.

**Comment by siuc-nate** [comment]
Sorry to jump in so late on this one - we just discovered an issue:

This request to the `/snhupr/` community is returning the copy of the framework from the default (`/ce-registry/`) community:
![image](https://user-images.githubusercontent.com/21346519/152234685-98f512b4-7e28-4093-a902-b5d64bd78da3.png)

We didn't spot this problem sooner because that framework is published to both the `ce-registry` **and** `snhupr communities, using the same CTIDs across the board.

The data returned should come from the community that matches the one in the request (the `snhupr` one in this case).

**Comment by excelsior** [comment]
@siuc-nate Did you have a chance to test these improvements on sandbox yet?

**Comment by siuc-nate** [comment]
It looks like this is working, thanks.
